src/model_server/lib/toolkit_function_response.cpp

```diff
@@ -0,0 +1,35 @@
+#include <model_server/lib/toolkit_function_response.hpp>
+#include <iostream>
+
+namespace turi {
+
+toolkit_function_response_future::toolkit_function_response_future(std::function<toolkit_function_response_type()> exec_function) :
+  m_info(new response_info) {
+
+    auto captured_info = this->m_info; 
+
+m_info->response_future = std::async(std::launch::async, 
+  [=]() {
+     captured_info->response = exec_function();
+     captured_info->is_completed = true;
+     return true;
+  });
+
+ ASSERT_TRUE(m_info->response_future.valid());
+}
+
+const toolkit_function_response_type& toolkit_function_response_future::wait() const {
+  ASSERT_TRUE(m_info != nullptr); 
+
+  if(!m_info->is_completed) { 
+     ASSERT_TRUE(m_info->response_future.valid()); 
+     m_info->response_future.wait(); 
```

Comment:
I think you can only use `future` without `is_completed` to tell whether it's completed or not.

Reply:
I'm not sure I get what you you're saying.  is_completed is a side effect of the function the future is managing, and may be valid before the future returns.

Reply:
let me clarify, the bool value `is_completed` can be stored into `future` to indicate the job is done successfully or not. If you are unlocked on future.wait(), them it means the job is done and you can retrieve its job status from future.get().

Reply:
The future isn't valid anymore after get() is called; once called, it can't be called again.  We want to be able to call this function multiple times if necessary, so storing it in our own state is better.